File: hearts-scala/src/main/scala/funar/hearts/Free.scala
package funar.hearts

import cats._
import cats.implicits._
import cats.data._
import scala.annotation.tailrec

enum Free[F[_], A] {
  case Pure[F[_], A](result: A) extends Free[F[_], A]
  case Impure[F[_], A](f: F[Free[F, A]]) extends Free[F[_], A]
}

object Free {
  implicit def freeMonad[F[_]](implicit fFunctor: Functor[F]): Monad[Free[F, _]] =
    new Monad[Free[F, _]] {
      def pure[A](value: A) = Pure[F, A](value)

      def flatMap[A, B](fa: Free[F, A])(k: A => Free[F, B]): Free[F, B] =
        fa match {
          case Pure(result) => k(result)
          case Impure(f) => Impure(fFunctor.map(f)(_.flatMap(k)))
        }

      def tailRecM[A, B](a: A)(next: A => Free[F, Either[A, B]]): Free[F, B] = {
        def loop(a: A): Free[F, B] =
          next(a).flatMap {
            case Left(a0) => tailRecM(a0)(next)
            case Right(b) => Pure(b)
          }

        loop(a)
      }
    }

}

File: mike/hearts/src/Free.hs
module Free where 

{-
data DB a
  = Get Key (Value -> DB a)
  | Put Key Value (() -> DB a)
                         ^^^^
  | Return a
    ^^^^^^^^
->

data DB' self =
  = Get Key (Value -> self)
  | Put Key Value (() -> self)


-}
data Free f a =
     Pure a
   | Impure (f (Free f a))

instance Functor f => Functor (Free f) where
  fmap f (Pure a) = Pure (f a)
  fmap f (Impure r) = Impure (fmap (fmap f) r)

instance Functor f => Applicative (Free f) where
  pure = Pure
  (Pure f) <*> aa = fmap f aa
  (Impure fa) <*> aa = Impure (fmap (<*> aa) fa)

instance Functor f => Monad (Free f) where
    return = Pure
    
    (Pure a) >>= k = k a
    (Impure r) >>= k = Impure (fmap (>>= k) r)

runFree :: Monad m => (f (Free f a) -> (Free f a -> m a) -> m a) -> Free f a -> m a
runFree _ (Pure result) = return result
runFree rrr (Impure command) = rrr command (runFree rrr)
